### PR TITLE
ARROW-4993: [C++] Add simple build configuration summary

### PR DESCRIPTION
**Pull request by bkietz**
The summary will be output to stdout and (by default) to a json file as well.

### CODE REVIEW

**Comment by bkietz** [comment]
stdout summary looks like this:

```
-- ---------------------------------------------------------------------
-- Arrow version:                                 0.14.0-SNAPSHOT
-- 
-- Build configuration summary:
--  Generator: Ninja
--  Build type: RELEASE
--  Source directory: /home/bkietz/arrow/cpp
--  Arrow options:
--   Only define the lint and check-format targets                          OFF [default, change using ARROW_ONLY_LINT]
--   Build static libraries                                                 ON  [ARROW_BUILD_STATIC]
--   Build shared libraries                                                 ON  [ARROW_BUILD_SHARED]
--   Run the test suite using valgrind --tool=memcheck                      OFF [default, change using ARROW_TEST_MEMCHECK]
--   Enable Address Sanitizer checks                                        OFF [default, change using ARROW_USE_ASAN]
--   Use ccache when compiling (if available)                               ON  [ARROW_USE_CCACHE]
--   Use ld.gold for linking on Linux (if available)                        OFF [default, change using ARROW_USE_LD_GOLD]
--   Enable Thread Sanitizer checks                                         OFF [default, change using ARROW_USE_TSAN]
--   Build the Arrow googletest unit tests, default OFF                     OFF [default, change using ARROW_BUILD_TESTS]
--   Build the Arrow micro benchmarks, default OFF                          OFF [default, change using ARROW_BUILD_BENCHMARKS]
--   Build the Arrow examples, default OFF                                  OFF [default, change using ARROW_BUILD_EXAMPLES]
--   Exclude deprecated APIs from build                                     OFF [default, change using ARROW_NO_DEPRECATED_API]
--   Build Arrow Fuzzing executables                                        OFF [default, change using ARROW_FUZZING]
--   Build with SIMD optimizations                                          ON  [ARROW_USE_SIMD]
--   Build Arrow with Altivec                                               ON  [ARROW_ALTIVEC]
--   Build Arrow commandline utilities                                      ON  [ARROW_BUILD_UTILITIES]
--   Build Arrow libraries with RATH set to $ORIGIN                         OFF [default, change using ARROW_RPATH_ORIGIN]
--   Build Arrow libraries with install_name set to @rpath                  ON  [ARROW_INSTALL_NAME_RPATH]
--   Build with C++ code coverage enabled                                   OFF [default, change using ARROW_GENERATE_COVERAGE]
--   If off, 'quiet' flags will be passed to linting tools                  OFF [default, change using ARROW_VERBOSE_LINT]
--   Pass -ggdb flag to debug builds                                        ON  [ARROW_GGDB_DEBUG]
--   Build the Arrow Compute Modules                                        ON  [ARROW_COMPUTE]
--   Build the Arrow Flight RPC System (requires GRPC, Protocol Buffers)    OFF [default, change using ARROW_FLIGHT]
--   Build the Gandiva libraries                                            OFF [default, change using ARROW_GANDIVA]
--   Build the Parquet libraries                                            OFF [default, change using ARROW_PARQUET]
--   Build the Arrow IPC extensions                                         ON  [ARROW_IPC]
--   Build the Arrow CUDA extensions (requires CUDA toolkit)                OFF [default, change using ARROW_CUDA]
--   Build the Arrow ORC adapter                                            OFF [default, change using ARROW_ORC]
--   Build Arrow with TensorFlow support enabled                            OFF [default, change using ARROW_TENSORFLOW]
--   Build the Arrow jemalloc-based allocator                               ON  [ARROW_JEMALLOC]
--   Build the Arrow HDFS bridge                                            ON  [ARROW_HDFS]
--   Build the Arrow CPython extensions                                     OFF [default, change using ARROW_PYTHON]
--   Build the HiveServer2 client and Arrow adapter                         OFF [default, change using ARROW_HIVESERVER2]
--   Build the plasma object store along with Arrow                         OFF [default, change using ARROW_PLASMA]
--   Build the plasma object store java client                              OFF [default, change using ARROW_PLASMA_JAVA_CLIENT]
--   Show output from ExternalProjects rather than just logging to files    OFF [default, change using ARROW_VERBOSE_THIRDPARTY_BUILD]
--   Rely on boost shared libraries where relevant                          ON  [ARROW_BOOST_USE_SHARED]
--   Use vendored Boost instead of existing Boost. Note that this requires linking Boost statically 
                                                                            OFF [default, change using ARROW_BOOST_VENDORED]
--   Rely on Protocol Buffers shared libraries where relevant               ON  [ARROW_PROTOBUF_USE_SHARED]
--   Rely on GFlags shared libraries where relevant                         ON  [ARROW_GFLAGS_USE_SHARED]
--   Build with backtrace support                                           ON  [ARROW_WITH_BACKTRACE]
--   Build libraries with glog support for pluggable logging                ON  [ARROW_USE_GLOG]
--   Build with Brotli compression                                          ON  [ARROW_WITH_BROTLI]
--   Build with BZ2 compression                                             OFF [default, change using ARROW_WITH_BZ2]
--   Build with lz4 compression                                             ON  [ARROW_WITH_LZ4]
--   Build with Snappy compression                                          ON  [ARROW_WITH_SNAPPY]
--   Build with zlib compression                                            ON  [ARROW_WITH_ZLIB]
--   Build with zstd compression                                            ON  [ARROW_WITH_ZSTD]
--   Depend only on Thirdparty headers to build libparquet. Always OFF if building binaries 
                                                                            OFF [default, change using PARQUET_MINIMAL_DEPENDENCY]
--   Build the Parquet executable CLI tools. Requires static libraries to be built. 
                                                                            OFF [default, change using PARQUET_BUILD_EXECUTABLES]
--   Build the Parquet examples. Requires static libraries to be built.     OFF [default, change using PARQUET_BUILD_EXAMPLES]
--   Build the Gandiva JNI wrappers                                         OFF [default, change using ARROW_GANDIVA_JAVA]
--   Include -static-libstdc++ -static-libgcc when linking with Gandiva static libraries 
                                                                            OFF [default, change using ARROW_GANDIVA_STATIC_LIBSTDCPP]
--   Compile with extra error context (line numbers, code)                  OFF [default, change using ARROW_EXTRA_ERROR_CONTEXT]
--   If enabled install ONLY targets that have already been built. Please be advised that if this is enabled 'install' will fail silently on components that have not been built 
                                                                            OFF [default, change using ARROW_OPTIONAL_INSTALL]
--  Outputting build configuration summary to /home/bkietz/arrow/cpp/build/test/cmake_summary.json
```

File: cpp/cmake_modules/DefineOptions.cmake [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+macro(define_option name description default)
+  option(${name} ${description} ${default})
+  list(APPEND ARROW_ALL_OPTION_NAMES ${name})
+  set("${name}_OPTION_DESCRIPTION" ${description})
+  set("${name}_OPTION_DEFAULT" ${default})
+endmacro()
+
+# Top level cmake dir
+if("${CMAKE_SOURCE_DIR}" STREQUAL "${CMAKE_CURRENT_SOURCE_DIR}")
+  set(ARROW_CXXFLAGS "" CACHE STRING "Compiler flags to append when compiling Arrow")
```
Comment (kou) [inline cpp/cmake_modules/DefineOptions.cmake]: Can we show no ON/OFF build options such as this?

**Review by kou — COMMENTED** [review]
Generally, I like it.
Could you check my comment?

**Comment by pitrou** [comment]
Is it easily doable to group those settings in categories? (at least in the display)

**Comment by bkietz** [comment]
@pitrou it's easy, which categories would you like to see?

File: cpp/cmake_modules/DefineOptions.cmake (hunk 1 above)
Comment (bkietz) [reply cpp/cmake_modules/DefineOptions.cmake]: @kou Yes, I can include `ARROW_CXXFLAGS` in the output. Are there any other variables you'd like to see?

Comment (bkietz) [reply cpp/cmake_modules/DefineOptions.cmake]: I'll add `ARROW_CXXFLAGS`, `ARROW_TEST_LINKAGE`, `ARROW_DEPENDENCY_SOURCE`, `ARROW_WITH_ZSTD`, `ARROW_GANDIVA_PC_CXX_FLAGS`, and [these msvc options](https://github.com/apache/arrow/pull/4305/files#diff-b048bf4c1679dce1028fd897a7c43b93R200)

**Comment by bkietz** [comment]
Current output:
```
-- ---------------------------------------------------------------------
-- Arrow version:                                 0.14.0-SNAPSHOT
--
-- Build configuration summary:
--   Generator: Unix Makefiles
--   Build type: RELEASE
--   Source directory: /home/bkietz/arrow/cpp
--   Outputting build configuration summary to /home/bkietz/arrow/cpp/build/test/cmake_summary.json
--
-- Compile and link options:
--   Compiler flags to append when compiling Arrow                          ""    [default] [ARROW_CXXFLAGS]
--   Build static libraries                                                 ON    [default] [ARROW_BUILD_STATIC]
--   Build shared libraries                                                 ON    [default] [ARROW_BUILD_SHARED]
--   Exclude deprecated APIs from build                                     OFF   [default] [ARROW_NO_DEPRECATED_API]
--   Use ccache when compiling (if available)                               ON    [default] [ARROW_USE_CCACHE]
--   Use ld.gold for linking on Linux (if available)                        OFF   [default] [ARROW_USE_LD_GOLD]
--   Build with SIMD optimizations                                          ON    [default] [ARROW_USE_SIMD]
--   Build Arrow with Altivec                                               ON    [default] [ARROW_ALTIVEC]
--   Build Arrow libraries with RATH set to $ORIGIN                         OFF   [default] [ARROW_RPATH_ORIGIN]
--   Build Arrow libraries with install_name set to @rpath                  ON    [default] [ARROW_INSTALL_NAME_RPATH]
--   Pass -ggdb flag to debug builds                                        ON    [default] [ARROW_GGDB_DEBUG]
--
-- Test and benchmark options:
--   Build the Arrow examples, default OFF                                  OFF      [default] [ARROW_BUILD_EXAMPLES]
--   Build the Arrow googletest unit tests, default OFF                     OFF      [default] [ARROW_BUILD_TESTS]
--   Build the Arrow micro benchmarks, default OFF                          OFF      [default] [ARROW_BUILD_BENCHMARKS]
--   Linkage of Arrow libraries with unit tests executables.                "shared" [default] [ARROW_TEST_LINKAGE] [shared|static]
--   Build Arrow Fuzzing executables                                        OFF      [default] [ARROW_FUZZING]
--
-- Lint options:
--   Only define the lint and check-format targets                          OFF   [default] [ARROW_ONLY_LINT]
--   If off, 'quiet' flags will be passed to linting tools                  OFF   [default] [ARROW_VERBOSE_LINT]
--   Build with C++ code coverage enabled                                   OFF   [default] [ARROW_GENERATE_COVERAGE]
--
-- Checks options:
--   Run the test suite using valgrind --tool=memcheck                      OFF   [default] [ARROW_TEST_MEMCHECK]
--   Enable Address Sanitizer checks                                        OFF   [default] [ARROW_USE_ASAN]
--   Enable Thread Sanitizer checks                                         OFF   [default] [ARROW_USE_TSAN]
--
-- Project component options:
--   Build the Arrow Compute Modules                                        ON    [default] [ARROW_COMPUTE]
--   Build the Arrow Flight RPC System (requires GRPC, Protocol Buffers)    OFF   [default] [ARROW_FLIGHT]
--   Build the Gandiva libraries                                            OFF   [default] [ARROW_GANDIVA]
--   Build the Parquet libraries                                            OFF   [default] [ARROW_PARQUET]
--   Build the Arrow IPC extensions                                         ON    [default] [ARROW_IPC]
--   Build Arrow commandline utilities                                      ON    [default] [ARROW_BUILD_UTILITIES]
--   Build the Arrow CUDA extensions (requires CUDA toolkit)                OFF   [default] [ARROW_CUDA]
--   Build the Arrow ORC adapter                                            OFF   [default] [ARROW_ORC]
--   Build Arrow with TensorFlow support enabled                            OFF   [default] [ARROW_TENSORFLOW]
--   Build the Arrow jemalloc-based allocator                               ON    [default] [ARROW_JEMALLOC]
--   Build the Arrow HDFS bridge                                            ON    [default] [ARROW_HDFS]
--   Build the Arrow CPython extensions                                     OFF   [default] [ARROW_PYTHON]
--   Build the HiveServer2 client and Arrow adapter                         OFF   [default] [ARROW_HIVESERVER2]
--   Build the plasma object store along with Arrow                         OFF   [default] [ARROW_PLASMA]
--   Build the plasma object store java client                              OFF   [default] [ARROW_PLASMA_JAVA_CLIENT]
--
-- Thirdparty toolchain options:
--   Method to use for acquiring arrow's build dependencies                 "CONDA" [default] [ARROW_DEPENDENCY_SOURCE] [AUTO|BUNDLED|SYSTEM|CONDA|BREW]
--   Show output from ExternalProjects rather than just logging to files    OFF     [default] [ARROW_VERBOSE_THIRDPARTY_BUILD]
--   Rely on boost shared libraries where relevant                          ON      [default] [ARROW_BOOST_USE_SHARED]
--   Use vendored Boost instead of existing Boost. Note that this requires linking Boost statically
                                                                            OFF     [default] [ARROW_BOOST_VENDORED]
--   Rely on Protocol Buffers shared libraries where relevant               ON      [default] [ARROW_PROTOBUF_USE_SHARED]
--   Rely on GFlags shared libraries where relevant                         ON      [default] [ARROW_GFLAGS_USE_SHARED]
--   Build with backtrace support                                           ON      [default] [ARROW_WITH_BACKTRACE]
--   Build libraries with glog support for pluggable logging                ON      [default] [ARROW_USE_GLOG]
--   Build with Brotli compression                                          ON      [default] [ARROW_WITH_BROTLI]
--   Build with BZ2 compression                                             OFF     [default] [ARROW_WITH_BZ2]
--   Build with lz4 compression                                             ON      [default] [ARROW_WITH_LZ4]
--   Build with Snappy compression                                          ON      [default] [ARROW_WITH_SNAPPY]
--   Build with zlib compression                                            ON      [default] [ARROW_WITH_ZLIB]
--   Build with zstd compression                                            ON      [default] [ARROW_WITH_ZSTD]
--
-- Parquet options:
--   Depend only on Thirdparty headers to build libparquet. Always OFF if building binaries
                                                                            OFF   [default] [PARQUET_MINIMAL_DEPENDENCY]
--   Build the Parquet executable CLI tools. Requires static libraries to be built.
                                                                            OFF   [default] [PARQUET_BUILD_EXECUTABLES]
--   Build the Parquet examples. Requires static libraries to be built.     OFF   [default] [PARQUET_BUILD_EXAMPLES]
--
-- Gandiva options:
--   Build the Gandiva JNI wrappers                                         OFF   [default] [ARROW_GANDIVA_JAVA]
--   Include -static-libstdc++ -static-libgcc when linking with Gandiva static libraries
                                                                            OFF   [default] [ARROW_GANDIVA_STATIC_LIBSTDCPP]
--   Compiler flags to append when pre-compiling Gandiva operations         ""    [default] [ARROW_GANDIVA_PC_CXX_FLAGS]
--
-- Advanced developer options:
--   Compile with extra error context (line numbers, code)                  OFF   [default] [ARROW_EXTRA_ERROR_CONTEXT]
--   If enabled install ONLY targets that have already been built. Please be advised that if this is enabled 'install' will fail silently on components that
have not been built
                                                                            OFF   [default] [ARROW_OPTIONAL_INSTALL]
```

**Comment by pitrou** [comment]
@bkietz Looks great!

File: cpp/cmake_modules/DefineOptions.cmake (hunk 1 above)
Comment (kou) [reply cpp/cmake_modules/DefineOptions.cmake]: Thanks!

**Review by kou — APPROVED** [review]
+1
We can merge this when we resolve https://travis-ci.org/apache/arrow/jobs/532481643#L2158-L2162

```text
CMake Error at cmake_modules/DefineOptions.cmake:39 (list):
  list does not recognize sub-command JOIN
Call Stack (most recent call first):
  cmake_modules/DefineOptions.cmake:50 (define_option_string)
  CMakeLists.txt:101 (include)
```

It seems that we can't use `list(JOIN ...)` because it's only supported by newer CMake.

File: cpp/cmake_modules/DefineOptions.cmake [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+
+macro(define_option name description default)
+  option(${name} ${description} ${default})
+  list(APPEND "ARROW_${ARROW_OPTION_CATEGORY}_OPTION_NAMES" ${name})
+  set("${name}_OPTION_DESCRIPTION" ${description})
+  set("${name}_OPTION_DEFAULT" ${default})
+  set("${name}_OPTION_TYPE" "bool")
+endmacro()
+
+function(join lst glue out)
```
Comment (kou) [inline cpp/cmake_modules/DefineOptions.cmake]: I feel that `join` is too generic name.
How about `list_join`?

Comment (kou) [reply cpp/cmake_modules/DefineOptions.cmake]: Or `join_list`?

**Pull request closed without merging**